### ISSUE TITLE
No code to inference

### ISSUE THREAD

**Issue by divyavani01** [BFeng14/ActFound#1]
I couldn't find the code to run model inference on my custom dataset to predict ic50 value in the repository....can someone help me out with this.

**Comment by BFeng14** [comment]
Thank you for your interest in our work, I will provide a convenient way for prediction on custom datasets soon, and you can follow this guideline before that:
1. please first define your own data loader function read_xxxx() following read_FEP_SET, and return {"ligand_sets": ligand_sets, "assays": list(ligand_sets.keys())}. ligand_sets is a dict, the keys are names for each assay, and values are list of ligand info which contains experimental pic50 and smiles.
https://github.com/BFeng14/ActFound/blob/2e414a8ea57b0f3686ab2079d496f0388065d120/dataset/load_dataset.py#L77

2. please modify the dataset class in the 95 line of data_chemblbdb_Assay_reg.py, add the following line:
```
elif self.args.expert_test == "xxxx":
    experiment_test = preprocess.read_xxxx()
```
https://github.com/BFeng14/ActFound/blob/2e414a8ea57b0f3686ab2079d496f0388065d120/dataset/data_chemblbdb_Assay_reg.py#L95

3. please download checkpoint files from Google Drive or figshare and unzip it. Then please run main_reg.py as follows, this will run the test using 0.2 to 0.8 fraction of data for finetuning, the finetuning ligands are selected uniformly random, and each test will be repeated 40 times:
```
CUSTOM_DATA="xxxx"
FIXED_PARAM="--test_sup_num [0.2,0.4,0.6,0.8] --test_repeat_num 40 --train 0 --test_epoch -1 --expert_test {CUSTOM_DATA}"
CHEMBL_DIR="./checkpoints_all/checkpoints_chembl"
CHEMBL_RES="./test_results/{CUSTOM_DATA}/chembl"
python main_reg.py --datasource=chembl --logdir ${CHEMBL_DIR}/checkpoint_chembl_actfound --model_name actfound --test_write_file ${CHEMBL_RES} ${FIXED_PARAM}
```

**Comment by divyavani01** [comment]
Thank you for your response,
I don't have experimental pic50 values or any actual values I have my own data and wanted to try out the predictions for my data using this model..
In this case I don't have y_task which should be my actual experimental values,so I have initially given all 0's and tried but later debugged the code and came to know y_task is much needed.

target_loss, target_preds = self.net_forward(x=x_task,
                                                                 y=y_task,
                                                                 assay_idx=assay_idx,
                                                                 split=split,
                                                                 weights=names_weights_copy,
                                                                 backup_running_statistics=False,
                                                                 training=True,
                                                                 num_step=num_step)
So can you help me out how to do this without actual_y values.

**Comment by BFeng14** [comment]
Sorry that Actfound is a meta learning model designed for few-shot cases, which means that we need at least a few ligands with known experimental activity of a certain assay to give accurate prediction on ligands without experimental activity. You can now turn to [data_pqsar_Assay_reg.py](https://github.com/BFeng14/ActFound/blob/main/dataset/data_pqsar_Assay_reg.py) to figure out how to do so (e.g. predict activity of 100 unknown activity ligands, using 10 ligands with known experimental activity from the inhibition of XXX target assay for finetuning). 

If you do want to predict activity without any ligands with known experimental activity for finetuning, you can just simply load the checkpoint and modify system_protonet.py to only output the result of self.regressor.forward. However, this return value don't have any exact meaning, and it can be regaurded as "the average activity over all biology assay".

For details of Actfound, please turn to our paper: https://doi.org/10.21203/rs.3.rs-3546062/v1
 
```
out = self.regressor.forward(x=x, split=split, params=weights,
                                     training=training,
                                     backup_running_statistics=backup_running_statistics,
                                     num_step=num_step)
```

**Comment by divyavani01** [comment]
I have a limited data for few assays that are inhibiting a particular target with their y_actual values...now I have another custom dataset for which i need predictions. So now I have used your model and for predicting them using the code you have provided. I don't have any y_actual values for my custom set...can u help out how to proceed further.

Previously you mentioned about self.regressor.forward() method, that method also taking y_task value, if I assign actual labels as 0s, the predict_all in net_forward() method is returning some unknown values..which are not related to our training activity values.

**Comment by BFeng14** [comment]
Sorry for my fault. Please write the init func of your learning system as follows
```
import torch
import torch.nn as nn
import numpy as np
import json
from .system_base import RegressorBase

from models.meta_neural_network_architectures import FCNReLUNormNetworkQSAR, AssayFCNReLUNormNetworkReg

class XXXXRegressor(RegressorBase):    
    def __init__(self, input_shape, args):
        super(XXXXRegressor, self).__init__(input_shape, args)
        self.regressor = FCNReLUNormNetworkQSAR(input_shape=self.input_shape, args=self.args, meta=True).cuda()
        self.softmax = nn.Softmax(dim=0)
        self.post_init(args)        
```

and get predicted value using this:
```
out_embed, out_value = self.regressor.forward(x=x, params=weights,
                                                      training=training,
                                                      backup_running_statistics=backup_running_statistics,
                                                      num_step=num_step)
```

Again, this value don't have any exact meaning. If you want to make use of your inhibition assays, you can train your own model using actfound as an initialization. We have done the same thing in the GDSC dataset, please refer to the paper for details, and the script for runing GDSC is given in scripts/[gdsc_test.sh](https://github.com/BFeng14/ActFound/blob/main/scripts/gdsc_test.sh), and the corresponding learning system is given in [system_gdsc.py](https://github.com/BFeng14/ActFound/blob/main/learning_system/system_gdsc.py).

**Comment by divyavani01** [comment]
Those inhibition assays picked from chembl .... now I have a small dataset which contains assaytypes for few smiles ....now how can i make use of actfound to predict my custom dataset of smiles for those assays or a few assays among the whole.

**Comment by BFeng14** [comment]
If there is no finetuning ligands (with experimental activity) provided, then we can only run self.regressor.forward to get the predicted value. Assay type (or assay taxonomy, assay description) is a good message for bio-activity prediction, however we didn't take them as input in the ActFound. It can be a interesting research project for us in the future.

For your custom test assay, if there happens to exist some similiar assays in the pretraining dataset (i.e. chembl), the predicted affinity may have good rank correlation with the real activity value. You can following the code I said above to get the prediction result. If you have problem in implementing this by yourself, I can help you with it in this weedend. Thank you for your interest in ActFound, please let me know if you have any other questions.
